controllers/clustergroupupgrade_controller.go

```diff
@@ -1197,33 +1193,22 @@ func (r *ClusterGroupUpgradeReconciler) getNextNonCompliantPolicyForCluster(
    returns: true/false if the upgrade is complete
             error/nil
 */
-func (r *ClusterGroupUpgradeReconciler) isUpgradeComplete(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) (bool, error) {
+func (r *ClusterGroupUpgradeReconciler) isUpgradeComplete(clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) bool {
 	// If we are not at the last batch, the upgrade clearly didn't complete.
-	if clusterGroupUpgrade.Status.Status.CurrentBatch < len(clusterGroupUpgrade.Status.CopiedPolicies) {
-		return false, nil
+	if clusterGroupUpgrade.Status.Status.CurrentBatch < len(clusterGroupUpgrade.Status.RemediationPlan) {
+		return false
 	}
 
-	// Go through all the clusters in the current last batch (which is also the last batch) and make sure
+	// Go through all the clusters in the current batch (which is also the last batch) and make sure
 	// they are either compliant with all the managedPolicies or they don't match them.
 	for _, batchClusterName := range clusterGroupUpgrade.Status.RemediationPlan[clusterGroupUpgrade.Status.Status.CurrentBatch-1] {
-		currentPolicyIndex := clusterGroupUpgrade.Status.Status.CurrentRemediationPolicyIndex[batchClusterName]
-		if currentPolicyIndex == utils.AllPoliciesValidated || currentPolicyIndex == utils.NoPolicyIndex {
-			continue
-		}
-
-		nextNonCompliantPolicyIndex, err := r.getNextNonCompliantPolicyForCluster(ctx, clusterGroupUpgrade, batchClusterName, currentPolicyIndex)
-		if err != nil {
-			return false, err
-		}
-
-		// If we can find a managed policy for which the cluster is NonCompliant, it means the upgrade has
-		// not finished.
-		if nextNonCompliantPolicyIndex != utils.NoPolicyIndex {
-			return false, nil
+		clusterProgress := clusterGroupUpgrade.Status.Status.CurrentBatchRemediationProgress[batchClusterName]
```

Comment:
We can chat about this. The point of the function is to check again periodically after it reaches the timeout state. Obviously it wasn't exactly working due to the bug you fixed a few lines above. However, it is mentioned in our readme. If we want to change the behavior, we should create a JIRA and do it in another PR.

Reply:
@jc-rh  I agree but I don't understand how the behaviour is changed. It seems to me that the function wants to check whether all the clusters in the current batch are completed. I think the changes are consistent with the past behaviour. It checks every cluster status to be completed.

Reply:
Maybe I missed something here? The original code actually checks the policy compliant status again. The new code only looks at the policy index map?

Reply:
you are right I missed it. I will fix it.

Reply:
@jc-rh I updated this part, although I am pretty sure that this part of the code doesn't work as expected currently.